src/Frame.cc

```diff
@@ -109,3 +169,121 @@ void Frame::ClearTrigger()
 	Unref(trigger);
 	trigger = 0;
 	}
+
+ClosureFrame::ClosureFrame(Frame* closure, Frame* not_closure,
+	std::shared_ptr<id_list> outer_ids) : Frame(not_closure)
+	{
+	assert(closure);
+
+	this->closure = closure;
+	Ref(this->closure);
+	this->body = not_closure;
+	Ref(this->body);
+
+	// To clone a ClosureFrame we null outer_ids and then copy
+	// the set over directly, hence the check.
+	if (outer_ids)
+		{
+		// Install the closure IDs
+		id_list* tmp = outer_ids.get();
+		loop_over_list(*tmp, i)
+			{
+			ID* id = (*tmp)[i];
+			if (id)
+				this->closure_elements.insert(id->Name());
+			}
+		}
+	}
+
+ClosureFrame::~ClosureFrame()
+	{
+	Unref(this->closure);
+	Unref(this->body);
+	}
+
+Val* ClosureFrame::GetElement(ID* id) const
+	{
+	if (this->closure_elements.find(id->Name()) != this->closure_elements.end())
+		return ClosureFrame::GatherFromClosure(this, id);
+
+	return this->NthElement(id->Offset());
+	}
+
+void ClosureFrame::SetElement(const ID* id, Val* v)
+	{
+	if (this->closure_elements.find(id->Name()) != this->closure_elements.end())
+	  ClosureFrame::SetInClosure(this, id, v);
+	else
+	  this->Frame::SetElement(id->Offset(), v);
+	}
+
+Frame* ClosureFrame::Clone()
+	{
+	Frame* new_closure = this->closure->Clone();
+	Frame* new_regular = this->body->Clone();
+
+	ClosureFrame* cf = new ClosureFrame(new_closure, new_regular, nullptr);
+	cf->closure_elements = this->closure_elements;
+	return cf;
+	}
+
+Frame* ClosureFrame::SelectiveClone(id_list* choose)
```

Comment:
This is where selective cloning takes place and the part of this that could probably use the most thought. As far as I can think (assuming that everything is implementing traverse properly) [`OuterIDBindingFinder`](https://github.com/zeek/zeek/blob/7cf623b2ff2288ea01059b2a5ed4712e112f0526/src/Var.cc#L405) should catch everything that we need to clone, but I'm not too familiar with the corners of Zeek. I'd appreciate a second brain here.

Reply:
Selective cloning also happens in `Frame::SelectiveClone`. This is the more interesting of the two methods though.